src/Features/CSharp/Portable/UsePatternMatching/CSharpAsAndNullCheckDiagnosticAnalyzer.Analyzer.cs

```diff
@@ -0,0 +1,379 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Diagnostics;
+using System.Threading;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.CSharp.UsePatternMatching
+{
+    internal partial class CSharpAsAndNullCheckDiagnosticAnalyzer
+    {
+        private readonly struct Analyzer
+        {
+            private readonly SemanticModel _semanticModel;
+            private readonly ILocalSymbol _localSymbol;
+            private readonly BinaryExpressionSyntax _comparison;
+            private readonly ExpressionSyntax _operand;
+            private readonly SyntaxNode _localStatement;
+            private readonly SyntaxNode _enclosingBlock;
+            private readonly CancellationToken _cancellationToken;
+
+            private Analyzer(
+                SemanticModel semanticModel,
+                ILocalSymbol localSymbol,
+                BinaryExpressionSyntax comparison,
+                ExpressionSyntax operand,
+                SyntaxNode localStatement,
+                SyntaxNode enclosingBlock,
+                CancellationToken cancellationToken)
+            {
+                Debug.Assert(semanticModel != null);
+                Debug.Assert(localSymbol != null);
+                Debug.Assert(comparison != null);
+                Debug.Assert(operand != null);
+                Debug.Assert(localStatement.IsKind(SyntaxKind.LocalDeclarationStatement));
+                Debug.Assert(enclosingBlock.IsKind(SyntaxKind.Block));
```

Comment:
💡 Consider changing the parameter and field type (`LocalDeclarationStatementSyntax` and `BlockSyntax`) if this is truly a requirement.

Reply:
This is just to avoid a wasteful cast. The node does not required to be of those types, but that's basically what we're passing here, hence the assertion.